Mathlib/Data/Nat/Digits/Lemmas.lean

```diff
@@ -319,4 +319,190 @@ theorem getD_digits (n i : ℕ) {b : ℕ} (h : 2 ≤ b) : (digits b n).getD i 0
+    · exact digits_lt_base hb hi
+    · linarith
+  · simp [ofDigits_digits]
+
+theorem injOn_ofDigits {b : ℕ} (hb : 1 < b) (l : ℕ) :
+    Set.InjOn (ofDigits b) {L : List ℕ | l = L.length ∧ ∀ x ∈ L, x < b} :=
+  fun _ _ _ _ h ↦ ofDigits_inj_of_len_eq hb (by aesop) (by aesop) (by aesop) h
+
+/--
+The map `L ↦ Nat.ofDigits b L` is bijection from the set of list of natural integers of length `l`
+with coefficients `< b` to the set of natural integers `< b ^ l`.
+-/
+theorem bijOn_ofDigits {b : ℕ} (hb : 1 < b) (l : ℕ) :
+    Set.BijOn (ofDigits b) {L : List ℕ | l = L.length ∧ ∀ x ∈ L, x < b} {n | n < b ^ l} :=
+  ⟨mapsTo_ofDigits hb l, injOn_ofDigits hb l, surjOn_ofDigits hb l⟩
+
 end Nat
+
+namespace List
+
+open Nat
+
+/--
+The set of list of natural integers of length `l` with coefficients `< b` as a `Finset`.
+This can be seen as the set of lists of length `l` of the digits in base `b` of
+the integers `< b ^ l`.
+Having this set as a `Finset` can helpful for some proofs.
+-/
+noncomputable abbrev fixedLengthDigits {b : ℕ} (hb : 1 < b) (l : ℕ) : Finset (List ℕ) := by
+  have : Fintype {L : List ℕ | l = L.length ∧ ∀ x ∈ L, x < b} :=
+    Fintype.ofInjective (Set.MapsTo.restrict _ _ _ (mapsTo_ofDigits hb l))
+      <| (Set.MapsTo.restrict_inj (mapsTo_ofDigits hb l)).mpr <| injOn_ofDigits hb l
+  exact {L : List ℕ | l = L.length ∧ ∀ x ∈ L, x < b}.toFinset
+
+/--
+The bijection `Nat.bijOn_ofDigits` stated as a bijection between `Finset`.
+This spelling can helpful for some proofs.
+-/
+theorem _root_.Nat.bijOn_ofDigits' {b : ℕ} (hb : 1 < b) (l : ℕ) :
+    Set.BijOn (ofDigits b) (fixedLengthDigits hb l) (Finset.range (b ^ l)) := by
+  simp only [Set.coe_toFinset]
+  convert bijOn_ofDigits hb l
+  ext; simp
+
+
+theorem mem_fixedLengthDigits_iff {b : ℕ} (hb : 1 < b) {l : ℕ} {L : List ℕ} :
+    L ∈ fixedLengthDigits hb l ↔ l = L.length ∧ ∀ x ∈ L, x < b := by
+  simp
+
+@[simp]
+theorem fixedLengthDigits_zero {b : ℕ} (hb : 1 < b) :
+    fixedLengthDigits hb 0 = {[]} := by
+  ext _
+  simpa [eq_comm] using by grind
+
+@[simp]
+theorem fixedLengthDigits_one {b : ℕ} (hb : 1 < b) :
+    fixedLengthDigits hb 1 = Finset.image (fun x : ℕ ↦ [x]) (Finset.range b) := by
+  ext
+  rw [mem_fixedLengthDigits_iff, eq_comm, List.length_eq_one_iff]
+  grind
+
+theorem fixedLengthDigits_card {b : ℕ} (hb : 1 < b) (l : ℕ) :
+    Finset.card (fixedLengthDigits hb l) = b ^ l := by
+  rw [Set.BijOn.finsetCard_eq (ofDigits b) (bijOn_ofDigits' hb l), Finset.card_range]
+
+/--
+Let `L` be a list in `List.fixedLengthDigits b l`. Then `L` is equal to the list of digits of
+`ofDigits b L` concatenated with enough `0`'s to that its length is equal to `l`.
+-/
+theorem eq_digits_ofDigits_append_of_mem_fixedLengthDigits {b : ℕ} (hb : 1 < b) (l : ℕ)
+    {L : List ℕ} (hL : L ∈ fixedLengthDigits hb l) :
+    L = b.digits (ofDigits b L) ++ replicate (l - (b.digits (ofDigits b L)).length) 0 := by
+  refine (Nat.bijOn_ofDigits' hb l).2.1 hL ?_ (by simp [Nat.ofDigits_digits])
```

Comment:
You're basically duplicating the `SurjOn` proof from above here, consider deduplicating it like this: https://github.com/TwoFX/mathlib4/commit/2080a46231557814848b2af33972132eb02c4ac9